### PR TITLE
Separate DirectionsWidget from MapWidget

### LINKED ISSUE #84: Show generated directions on the map
Use polylines to show the generated directions on the map.

Acceptance Criteria: 
- [ ] Once an itinerary is selected, there is also a drawn path showing the directions the user must follow to reach the destination.

### LINKED ISSUE #100: Calling length on null during first call to get directions
After initial startup of the app when getting the list of directions, the json object is null resulting in trying to get the length of a null object.

Reproduction steps:
1. Select search icon
1. Select a destination

### LINKED ISSUE #104: Separate Map and Directions Widgets
#98 adds directions to an [openable/closable widget](https://github.com/MichelRahme/ConcordiaNavigation/blob/feature/98-directions-list/lib/widgets/directions_drawer.dart) below the map but introduces a problem whereby said widget is refreshed along with the map every 2-3 seconds (firstly introduced in 97bce0daef0d92e50d24d133fd3f9bf7dd35b739)

This is done via [the call](https://github.com/MichelRahme/ConcordiaNavigation/blob/feature/98-directions-list/lib/services/location_service.dart#L57) to [onLocationChanged](https://github.com/Lyokone/flutterlocation/blob/master/lib/location.dart#L196) from [package:location/location.dart](https://pub.dev/packages/location)

Updating the location is useful if we want the map to follow the user. We have to find a way to prevent the rebuild of the direction widget, while still rebuilding the map.

One possible solution would be to convert the DirectionsDrawer to a BottomSheet Widget in order to have it as the `bottomSheet: ` property of the app's Scaffold element, giving the following result:

screens/home_page.dart
```dart
...
return Scaffold(
  appBar: HomePageAppBar(),
  drawer: CustomDrawer(),
  body: MapWidget(),
  bottomSheet: DirectionsDrawer(),
);
...
```

This way, the map could update without affecting the DirectionsDrawer widget.

### PR DESCRIPTION
**Pull request by andrelegault**
# Description
This pull request separates DirectionsDrawer (`/lib/widgets/directions_widget.dart`) into its own widget and fixes a bug related to calling length on a null object.

The widget is shown whenever an itinerary is selected by the user, and can be closed by clicking on the 'x', or brought down the screen to show the map.

## Related issues
- Fixes #104 
- Fixes #100

### CODE REVIEW

File: lib/providers/shuttle_data.dart [hunk 1]
```diff
@@ -44,15 +46,15 @@ class ShuttleData extends ChangeNotifier {
...
-      "930",
-      "945",
+      "0730",
+      "0740",
+      "0755",
+      "0820",
+      "0835",
+      "0855",
+      "0910",
+      "0930",
```
Comment (WindOfRussia) [inline lib/providers/shuttle_data.dart]: Yeah a bit redundant but yeah we should refactor this our of the provider. #105

File: lib/providers/map_data.dart [hunk 2]
```diff
@@ -1,26 +1,24 @@
...
 import 'package:flutter/foundation.dart';
 import 'package:google_maps_flutter/google_maps_flutter.dart';
 import 'package:flutter/material.dart';
+import 'package:sliding_up_panel/sliding_up_panel.dart';
 
 ///Observer Pattern
 ///Handles all the data related to the map, listens to changes and notifies listeners.
 class MapData extends ChangeNotifier {
   Completer<GoogleMapController> _completer = Completer();
+  PanelController panelController = new PanelController();
```
Comment (WindOfRussia) [inline lib/providers/map_data.dart]: Just as an fyi for me, whats a panel controller?

File: lib/screens/home_page.dart [hunk 3]
```diff
@@ -15,7 +16,12 @@ class _HomePageState extends State<HomePage> {
     return Scaffold(
       appBar: HomePageAppBar(),
       drawer: CustomDrawer(),
-      body: MapWidget(),
+      body: Stack(
+        children: <Widget>[
+          MapWidget(),
```
Comment (WindOfRussia) [inline lib/screens/home_page.dart]: Looks like mapwidget might be getting a bit too crowded again. Might be worth separating some items and extracting some widgets out of it.

File: lib/widgets/directions_drawer.dart [hunk 4]
```diff
@@ -1,51 +1,420 @@
...
+                                              fontSize: 15.0,
+                                              fontWeight: FontWeight.w400,
+                                              color: constants.blackColor,
+                                            ),
+                                          ),
+                                        ),
+                                      );
+                                    },
+                                  ),
+                                );
```
Comment (WindOfRussia) [inline lib/widgets/directions_drawer.dart]: Same comment as before, we should look into extracting some widgets out of this drawer and into their own files.

File: lib/providers/map_data.dart (hunk 2 above)
Comment (andrelegault) [reply lib/providers/map_data.dart]: The panel controller is the controller class for the SlideUpPanel widget that contains the DirectionsDrawer.

With it, you can do things like open or hide the panel. I put it in the provider to make it accessible throughout the app. [here's the docs for it](https://pub.dev/documentation/sliding_up_panel/latest/sliding_up_panel/PanelController-class.html).

This way, when an element wants to close the directions widget (it is pretty obtrusive) they can call `Provider.of<MapData>(context, listen: false).panelController.show()`.

File: lib/providers/map_data.dart [hunk 5]
```diff
@@ -98,8 +78,11 @@ class MapData extends ChangeNotifier {
     notifyListeners();
   }
 
-  final controllerStarting = TextEditingController();
-  final controllerDestination = TextEditingController();
+  void removeItinerary() {
```
Comment (MichelRahme) [inline lib/providers/map_data.dart]: can you specify what this function is doing? Is it responsible of hiding the directions panel? if yes, please specify it in a comment or make the function name more informative.

File: lib/screens/home_page.dart [hunk 6]
```diff
@@ -15,7 +16,12 @@ class _HomePageState extends State<HomePage> {
     return Scaffold(
       appBar: HomePageAppBar(),
       drawer: CustomDrawer(),
-      body: MapWidget(),
+      body: Stack(
```
Comment (MichelRahme) [inline lib/screens/home_page.dart]: I think the stack is not needed here. Can you please verify?

File: lib/widgets/directions_drawer.dart [hunk 7]
```diff
@@ -1,51 +1,420 @@
...
+    Future<Map<String, Map<String, String>>> _fetchMoreData() async {
+      Map<String, Map<String, String>> test =
+          Provider.of<MapData>(context, listen: false)?.itinerary?.itinerary;
+      await Future.delayed(Duration(seconds: 1));
+      return Future.value(test);
+    }
+
+    SlidingUpPanel sp = SlidingUpPanel(
+        controller: Provider.of<MapData>(context).panelController,
+        maxHeight: 685.0,
```
Comment (MichelRahme) [inline lib/widgets/directions_drawer.dart]: Use `SizeConfig` for different screen sizes

File: lib/widgets/directions_drawer.dart [hunk 8]
```diff
@@ -1,51 +1,420 @@
...
+                                  ),
+                                );
+                            }
+                          }),
+                    ), // list of directions
+                    ShuttleWidget(),
+                  ],
+                )));
+
+    if (Provider.of<MapData>(context).itinerary == null) {
```
Comment (MichelRahme) [inline lib/widgets/directions_drawer.dart]: What's the purpose of this? Why is it here?

File: lib/screens/home_page.dart (hunk 6 above)
Comment (andrelegault) [reply lib/screens/home_page.dart]: I'm not sure what I should replace it with.

File: lib/widgets/directions_drawer.dart (hunk 8 above)
Comment (andrelegault) [reply lib/widgets/directions_drawer.dart]: Check 7c8a951

File: lib/widgets/directions_drawer.dart (hunk 7 above)
Comment (andrelegault) [reply lib/widgets/directions_drawer.dart]: Check 735bb9d

File: lib/providers/map_data.dart (hunk 5 above)
Comment (andrelegault) [reply lib/providers/map_data.dart]: Check 
4b36668

File: lib/screens/home_page.dart (hunk 6 above)
Comment (MichelRahme) [reply lib/screens/home_page.dart]: I reviewed the drawer you have implemented and indeed it requires a Stack. Thanks

**Comment by WindOfRussia** [comment]
All comments resolved. good to go.

**Pull request merged**